test/studies/shootout/fannkuch-redux/bradc/fannkuch-redux-blc.chpl

```diff
@@ -1,78 +1,119 @@
 /* The Computer Language Benchmarks Game
- * http://benchmarksgame.alioth.debian.org/
- *
- * contributed by Kyle Brady and Brad Chamberlain
- * based on C implementation by Ledrug Katz
- *
- */
+   http://benchmarksgame.alioth.debian.org/
 
-config const N = 7;
+   contributed by Ben Harshbarger and Brad Chamberlain
+   derived from the Swift implementation by Ralph Ganszky
+*/
 
-assert(N >=3, "N must be in 3..");
+use DynamicIters;
+
+config const n = 7,          // the array size over which to compute perms
+             nchunks = 720;  // the number of chunks of parallelism
+
+//
+// memoize n! (n factorial)
+//
+var fact: [0..n] int;
+fact[0] = 1;
+for i in 1..n do
+  fact[i] = i*fact[i-1];
 
-const D = {0..N};
-var s, t: [D] int;
 
 proc main() {
-  for i in D do     // TODO: Could use s = D here
-    s[i] = i;
-  const (checksum, maxFlips) = tk();
-  writeln(checksum);
-  writeln("Pfannkuchen(", N, ") = ", maxFlips);
+  var checkSum = 0,
+      maxFlips = 1;
+
+  var chunksz = (fact(n) + nchunks - 1) / nchunks;
+  chunksz += chunksz%2;
+  const work = 0..(fact(n) - chunksz) by chunksz;
+
+  forall lo in dynamic(work, 1) with (+ reduce checkSum, 
+                                      max reduce maxFlips) {
+    for (i, flips) in fannkuch(lo..#chunksz) {
+      maxFlips = max(maxFlips, flips);
+      checkSum += if i % 2 == 0 then flips else -flips;
+    }
+  }
+  
+  writeln(checkSum);
+  writeln("Pfannkuchen(", n, ") = ", maxFlips);
 }
 
 
-proc tk() {
-  var checksum, maxFlips = 0;
-  var odd = false;
-  var c: [D] int;
+iter fannkuch(inds) {
+  var p, pp, count: [0..#n] int;
+  p = 0..#n;
 
-  var i = 0;
-  while i < N {
-    rotate(i);
-    if c[i] >= i {
-      c[i] = 0;
-      i += 1;
-      continue;
-    }
+  firstPerm();
 
-    c[i] += 1;
-    i = 1;
-    odd = !odd;
-    if s[0] {
-      const f = if s[s[0]] then flip() else 1;
-      if f > maxFlips then maxFlips = f;
-      checksum += if odd then -f else f;
-    }
+  for i in inds {
+    if p[0] != 0 then
+      yield (i, countFlips());
+    
+    if i != inds.high then
+      nextPerm();
   }
 
-  return (checksum, maxFlips);
-}
 
+  proc firstPerm() {
+    var idx = inds.low;
+    for i in 1..n-1 by -1 {
+      const fact_i = fact(i),
+            d = idx / fact_i;
+      count[i] = d;
+      idx %= fact(i);
```

Comment:
Probably.  Though I'll also look into the perf impact of caching it vs. indexing it twice.